### PR TITLE
Refer to types using the local identifier

### LINKED ISSUE #21934: The compiler should report publicly exported type names if possible
If a private type is reexported at some public location, the Rust compiler will report the private location when referring to the type. It'd be more helpful if the compiler reported the public alias for it.

This was reported by Coda Hale on Twitter.

### PR DESCRIPTION
**Pull request by estebank**
On type errors, refer to types using the locally available name when
they have been imported into scope instead of the fully qualified path.

```
error[E0308]: mismatched types
 --> file.rs:7:24
  |
7 |     let y: Option<X> = Ok(());
  |                        ^^^^^^ expected enum `std::option::Option`, found enum `std::result::Result`
  |
  = note: expected type `Option<X>`
             found type `Result<(), _>`
  = help: here are some functions which might fulfill your needs:
          - .err()
          - .unwrap_err()
```

Re: #21934.

### CODE REVIEW

File: src/librustc/ty/context.rs [hunk 1]
```diff
@@ -1179,6 +1179,10 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         self.all_crate_nums(LOCAL_CRATE)
     }
 
+    pub fn cstore(&self) -> &CrateStore {
```
Comment (arielb1) [inline src/librustc/ty/context.rs]: I think @alexcrichton removed this for a reason.

File: src/librustc/infer/error_reporting/mod.rs [hunk 2]
```diff
@@ -357,7 +359,7 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                 // for imported and non-imported crates
                 if exp_path == found_path
                 || exp_abs_path == found_abs_path {
-                    let crate_name = self.tcx.crate_name(did1.krate);
+                    let crate_name = self.tcx.cstore().crate_name_untracked(did1.krate);
```
Comment (arielb1) [inline src/librustc/infer/error_reporting/mod.rs]: Why can't you use `crate_name` here?

File: src/librustc/infer/error_reporting/mod.rs [hunk 3]
```diff
@@ -966,3 +712,515 @@ impl<'tcx> ObligationCause<'tcx> {
...
 }
+
+/// Visitor that collects every `Item` imported into the current `TyCtxt`'s scope.
+struct ImportVisitor<'a, 'gcx: 'a + 'tcx, 'tcx: 'a> {
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    /// All the `Item`s from other scopes visible from this `TyCtxt`, as well as their local name.
+    full_imports: FxHashMap<String, String>,
+}
+
+impl<'a, 'gcx, 'tcx> ImportVisitor<'a, 'gcx, 'tcx> {
```
Comment (arielb1) [inline src/librustc/infer/error_reporting/mod.rs]: This looks likely to be buggy - you are reimplementing half of name resolution. Could name resolution be made to supply a `types_in_scope` map?

Comment (arielb1) [reply src/librustc/infer/error_reporting/mod.rs]: cc @jseyfried - is there a way for resolution to be called after it is complete?

File: src/librustc/infer/error_reporting/mod.rs [hunk 4]
```diff
@@ -966,3 +712,515 @@ impl<'tcx> ObligationCause<'tcx> {
         }
     }
 }
+
+/// Visitor that collects every `Item` imported into the current `TyCtxt`'s scope.
+struct ImportVisitor<'a, 'gcx: 'a + 'tcx, 'tcx: 'a> {
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    /// All the `Item`s from other scopes visible from this `TyCtxt`, as well as their local name.
+    full_imports: FxHashMap<String, String>,
```
Comment (arielb1) [inline src/librustc/infer/error_reporting/mod.rs]: Why not an  `FxHashMap<DefId, String>`?

File: src/librustc/infer/error_reporting/mod.rs [hunk 5]
```diff
@@ -966,3 +712,515 @@ impl<'tcx> ObligationCause<'tcx> {
...
+                        }
+                    }
+                }
+            }
+            _ => (), // Ignore non-import items
+        }
+    }
+}
+
+/// This struct knows how to print out a single type for human consumption, compare two types and
```
Comment (arielb1) [inline src/librustc/infer/error_reporting/mod.rs]: Can this replace ppaux?

File: src/librustc/infer/error_reporting/mod.rs [hunk 6]
```diff
@@ -966,3 +712,515 @@ impl<'tcx> ObligationCause<'tcx> {
...
+                if !r.is_empty() {
+                    s.push_str(" ");
+                }
+                if let hir::Mutability::MutMutable = ty_and_mut.mutbl {
+                    s.push_str("mut ")
+                }
+                s.push_str(&self.ty_str(ty_and_mut.ty));
+                s
+            }
+            _ => format!("{}", ty),
```
Comment (arielb1) [inline src/librustc/infer/error_reporting/mod.rs]: You really want to replace the entirety of ppaux here. Otherwise e.g. everything within a trait object will go back to ugly.

File: src/librustc/infer/error_reporting/mod.rs [hunk 7]
```diff
@@ -966,3 +712,515 @@ impl<'tcx> ObligationCause<'tcx> {
...
+                                            }
+                                            _ => (),  // item not visible, don't alias
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    } else {
+                        // Importing from a different crate.
+                        for child in self.tcx.cstore().item_children_untracked(def_id,
```
Comment (alexcrichton) [inline src/librustc/infer/error_reporting/mod.rs]: Ah so the "untracked" here means that it's not tracked by incremental compilation, so this is something that we're going to want to handle as this is all happening after the tcx is created. If you need help turning this into a query though just let me know!

File: src/librustc/infer/error_reporting/mod.rs (hunk 2 above)
Comment (estebank) [reply src/librustc/infer/error_reporting/mod.rs]: This is a spurious change that should not have happened. I introduced it after hastily rebasing to HEAD of master when the moving of `cstore` from `sess` to `tcx` happened. I'll revert.

File: src/librustc/infer/error_reporting/mod.rs (hunk 4 above)
Comment (estebank) [reply src/librustc/infer/error_reporting/mod.rs]: This is an artifact from the exploratory way I wrote this code, as while working on this I didn't have access to `DefId`s from other crates. Very sensible suggestion.

File: src/librustc/infer/error_reporting/mod.rs (hunk 5 above)
Comment (estebank) [reply src/librustc/infer/error_reporting/mod.rs]: That is my eventual objective. I have moved the type argument elision code away from `TyCtxt` in order to isolate as much from it as possible in order to unify it with the code in `ppaux`. Do you think I should strive to do it in this PR?

File: src/librustc/infer/error_reporting/mod.rs (hunk 6 above)
Comment (estebank) [reply src/librustc/infer/error_reporting/mod.rs]: Yes, that is indeed true. Given how big this PR was getting I thought that it was a good idea to put it out there to get feedback before going further down this rabbit hole. That being said, this covered a big enough percentage of cases that allowed me to test this approach.

File: src/librustc/ty/context.rs (hunk 1 above)
Comment (estebank) [reply src/librustc/ty/context.rs]: I'm sure. This was a quick fix after rebasing against master to get the code _as I had written it_ working. I fully intend to clean this up before attempting to merge this PR.

File: src/test/ui/local-ident.stderr [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+   |                ^ expected struct `Mod2::X`, found struct `Mod1::S`
+   |
+   = note: expected type `X`
+              found type `S`
+
+error[E0308]: mismatched types
+  --> $DIR/local-ident.rs:16:28
+   |
+16 |     let y: Option<usize> = Ok(2);
+   |                            ^^^^^ expected enum `std::option::Option`, found enum `std::result::Result`
```
Comment (estebank) [inline src/test/ui/local-ident.stderr]: This is an open question: Do we want to show the fully qualified path here? I don't think length is as much of an issue here as this label doesn't print the type arguments, keeping the line relatively short. Alternatively it might be jarring to have two different things being shown here...

File: src/test/ui/local-ident.stderr [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+   |
+   = note: expected type `Option<usize>`
+              found type `Result<{integer}, _>`
```
Comment (estebank) [inline src/test/ui/local-ident.stderr]: Should we add a `note` here stating `Option: std::option::Option` and `Result: std::result::Result`? This could become a very long list for cases like the one put forward in #40186.

File: src/librustc/infer/error_reporting/mod.rs (hunk 7 above)
Comment (estebank) [reply src/librustc/infer/error_reporting/mod.rs]: Could you point me in the right direction of a query being performed? I haven't worked in that part of the compiler yet and would appreciate having a thread to pull from to unravel that sweater.

File: src/librustc/infer/error_reporting/mod.rs (hunk 3 above)
Comment (estebank) [reply src/librustc/infer/error_reporting/mod.rs]: I figured that was the case, but it was the closest I've been in any of my approaches to get this working. Once I saw that the suggestion machinery used a visitor it gave me enough context to get this. Having this also allowed me to identify the unit tests that would need to change. It is definitely less than ideal and fully expect there to be a way to do this using existing datastructures, but I feel it is much easier for me to point at this code and saw "this is what I want to do, is there a better way?". :)

**Review by estebank — COMMENTED** [review]
TL;DR: Some of the weird code is the result of hastily rebasing against master and pushing the code out without reworking it to fit the current state of the compiler, and some is the result of the exploratory way I was writing the code without going back to clean it up after the reasons for the current state were no longer there.

File: src/librustc/infer/error_reporting/mod.rs (hunk 7 above)
Comment (alexcrichton) [reply src/librustc/infer/error_reporting/mod.rs]: Sure! You can check out https://github.com/rust-lang/rust/issues/41417 for some more info

**Comment by krdln** [comment]
What about shadowing? For example:

```rust
struct Result;

fn main() {
    let res: Result = Ok(());
}
```

This PR produces the following note:

```
   = note: expected type `Result`
              found type `Result<(), _>`
```

(the full path for `std::result::Result` is present in the former part of the error message though)

**Comment by alexcrichton** [comment]
ping @estebank just want to make sure this doesn't fall off your radar!

**Comment by nikomatsakis** [comment]
r? @arielb1 

Reassigning to @arielb1 since he did initial review.

File: src/librustc/infer/error_reporting/mod.rs (hunk 3 above)
Comment (jseyfried) [reply src/librustc/infer/error_reporting/mod.rs]: There's not a way to do name resolution after lowering to HIR (i.e. after we drop the `Resolver`). A naive solution would be to keep the `Resolver` for longer, but that would hog memory and isn't ideal since the Resolver has needless complexity used for import resolution and is based on the AST.

@nikomatsakis and I have discussed creating and saving some name resolution structures for this purpose before destroying the `Resolver`, but we haven't made detailed designs.

**Comment by arielb1** [comment]
Still waiting on resolution coordination with @jseyfried.

**Comment by aidanhs** [comment]
Not clear what the status here is - is there an issue that triagers can visit to see progress of any blockers? Just wondering who's 'responsible' for pushing this along :)

**Comment by aidanhs** [comment]
@arielb1 / @jseyfried / @estebank - can anyone clarify what/who this PR is waiting on at the moment (aside from a rebase)? I'm not sure how to triage this at the moment as I don't know what's blocking or if there's an issue we can track!

**Comment by estebank** [comment]
@aidanhs I need to rework the PR to use queries, but I haven't had time available to do so. If people prefer, I can close this PR and resubmit once I've reworked it.

**Comment by aidanhs** [comment]
@estebank ok great, thanks for clarifying! If you reckon you might get to this in the next couple of weeks then I'd be fine with keeping it open - up to you.

**Comment by alexcrichton** [comment]
Here for triage again! I'm going to go ahead and close this now to help clear out the queue, but @estebank of course feel free to resubmit once it's reworked!
